### ISSUE TITLE
spec3 drizzle call suddenly using MUCH more memory (60x for one example)

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6479]
_Issue [JP-2340](https://jira.stsci.edu/browse/JP-2340) was created on JIRA by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam):_

SDP noticed a new LEVEL_3 failure (just one, but we don’t run many cases) in our nightly AUTO regression tests under the new SCSB code (pulled Friday, 11/26/2021). It was a “memory failure” error and it appears that for a simple spec3 association (just 2 members, the NRS1,NRS2 chips) the memory demand went from just shy of 1 Gb to >60Gb (!!) :
 
AUTO runDate                                  LEVEL_3 memory footprint (Kbytes)
Nov 20 2021 07:30:36:000AM         939160
Nov 21 2021 07:18:33:000AM         940420
Nov 23 2021 07:23:24:000AM         917348
Nov 24 2021 07:14:34:000AM         939208
Nov 25 2021 07:15:51:000AM         937744
Nov 26 2021 07:16:16:000AM         923268
Nov 27 2021 07:18:54:000AM         61717252.  << when the SCSB release went in
Nov 28 2021 07:23:34:000AM         64010764
Nov 29 2021 07:03:50:000AM         61716972
 
The dataset is here, if that helps:
   /ifs/archive/dev/jwst/store/dmsauto/Nightly_tests/test_dmsauto_20211129_01/regr_10032/pipe/sdp/asn_creation/cal/level3/
       jw84700-o001_20211129t070029_spec3_002_asn.json

and the log is here:
    /ifs/archive/dev/jwst/store/dmsauto/Nightly_tests/test_dmsauto_20211129_01/regr_10032/output/level_3_jw84700-o001_20211129t070029_spec3_002_dmsauto_1638169424.287002/ ALOG_1638169429_level_3_jw84700-o001_20211129t070029_spec3_002.out

Can someone look into why such a huge memory jump would have occurred for such a case?  It looks like it was during the drizzle step.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564424#comment-564424):_

Is it possible to determine the Cal release/version numbers from just before and after the change in memory usage?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564427#comment-564427):_

Should be, yes.  Let me check and update this comment with that info.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564428#comment-564428):_

The huge difference between processing on 11/26 and 11/27 is that on 11/26 the output resampled image had a size of only 3268x35, which seems reasonable for NRS fixed-slit spectra taken in the S200A1 subarray. But ever since 11/27 the output resampled image size has been drastically larger, at 3268x132283! So that certainly explains the huge increase in memory usage by the resample_spec step. Now the question is what led to the huge change in the computed output image size. Changes in WCS info, changes in resample code, ...

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564496#comment-564496):_

The only real difference between the old and new processing runs is the value of the ROLL_REF keyword, which is computed by the level-1b set_telescope_pointing script. Previously, the individual exposures had ROLL_REF=359.7856701550469, which is essentially zero. When run through calwebb_spec3, the output combined image size is computed to be 3268x35. The newer processing results, however, have ROLL_REF=-138.7605896, which results in the output image size of 3268x132283. I have verified that simply setting ROLL_REF back to its old value (and rerunning earlier pipeline stages to rebuild the gwcs object) results in old, smaller output image size.

On a hunch that the negative ROLL_REF value might be causing issues somewhere, I added 360 to -138.7605896, which results in ROLL_REF=221.2394104, and tried processing with this new value. The output image size in calwebb_spec3 still comes out with the ridiculous values of 3268x132283. So the negative value does not appear to be the problem. The problem appears to be related to the fact that ROLL_REF is not near zero.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564501#comment-564501):_

I note that the value of ROLL_REF=-138.76 does seem to make sense, given that (for these exposures) PA_V3=0 and V3I_YANG=138.76, wth VPARITY=-1. So the roll at the aperture ref point is simply equal to the (negative of) the roll of the NRS focal plane in the V2/V3 frame.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564530#comment-564530):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) If either of you want to try to look into this, example datasets are in ███████████████████████████████████████████ There are 3 sets of files: "old" refers to the exposures using the old ROLL_REF value of 359.8, "new" refers to the exposures using the new ROLL_REF value of -138.8, and "new2" refers to the exposures using the new ROLL_REF value that's had 360 added to it (to make it positive 221.2). The asn files "jw84700-o001_20211129t070029_spec3_002_asn_[old|new|new2].json" can be used to run the calwebb_spec3 pipeline using the old, new, new2 exposures as input. When you run:

strun calwebb_spec3 <asn_file.json>

you can see in the outlier_detection step already that it's creating a median image of ridiculous size when processing either the "new" or "new2" exposures.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564551#comment-564551):_

Would getting this data make my internet provider charge me as a business?

**Comment by mcara** [comment]
I believe there are a couple of things that do not seem to work well in computing output pixel size for the spectral case.

First, the comment does not agree with the `if` statement:
https://github.com/spacetelescope/jwst/blob/57becc82a683fa65bda78ae9dfba08ca49de6696/jwst/resample/resample_spec.py#L248-L260

It seems to me it should be checking for `pix_to_xtan.slope` and not the `pix_to_ytan.slope`. Or, maybe the comment is just wrong. However, since x-slope is used to compute `x_size` (below), it is clear that a near 0 x-slope poses a problem.

Then, in this code:

https://github.com/spacetelescope/jwst/blob/57becc82a683fa65bda78ae9dfba08ca49de6696/jwst/resample/resample_spec.py#L285-L293

I believe in this code we should also swap `x<->y` when x-slope is near 0 but I am not sure (since I do not know how things are intended to work here). I will investigate more. However, a simple test shows that if we do that, the new size is more reasonable. However, the code now crashes because of bad inverse mapping computed in

https://github.com/spacetelescope/jwst/blob/57becc82a683fa65bda78ae9dfba08ca49de6696/jwst/resample/resample_spec.py#L255-L260

The error is:

```
astropy.modeling.core.ModelDefinitionError: Unsupported operands for |: None (n_inputs=3, n_outputs=3) and None (n_inputs=2, n_outputs=2); n_outputs for the left-hand model must match n_inputs for the right-hand model.
```

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564624#comment-564624):_

I proposed a fix in #6533

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564974#comment-564974):_

Fixed by #6533

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564424#comment-564424):_

Is it possible to determine the Cal release/version numbers from just before and after the change in memory usage?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564427#comment-564427):_

Should be, yes.  Let me check and update this comment with that info.

From the 2021-11-26 run that had lower-memory usage we see:
CAL_VER = '1.3.4.dev48+gec75e7b'

From the 2021-11-27 run that jumped in memory usage we see:
CAL_VER = '1.3.4.dev50+g421a39c'

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564428#comment-564428):_

The huge difference between processing on 11/26 and 11/27 is that on 11/26 the output resampled image had a size of only 3268x35, which seems reasonable for NRS fixed-slit spectra taken in the S200A1 subarray. But ever since 11/27 the output resampled image size has been drastically larger, at 3268x132283! So that certainly explains the huge increase in memory usage by the resample_spec step. Now the question is what led to the huge change in the computed output image size. Changes in WCS info, changes in resample code, ...

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564496#comment-564496):_

The only real difference between the old and new processing runs is the value of the ROLL_REF keyword, which is computed by the level-1b set_telescope_pointing script. Previously, the individual exposures had ROLL_REF=359.7856701550469, which is essentially zero. When run through calwebb_spec3, the output combined image size is computed to be 3268x35. The newer processing results, however, have ROLL_REF=-138.7605896, which results in the output image size of 3268x132283. I have verified that simply setting ROLL_REF back to its old value (and rerunning earlier pipeline stages to rebuild the gwcs object) results in old, smaller output image size.

On a hunch that the negative ROLL_REF value might be causing issues somewhere, I added 360 to -138.7605896, which results in ROLL_REF=221.2394104, and tried processing with this new value. The output image size in calwebb_spec3 still comes out with the ridiculous values of 3268x132283. So the negative value does not appear to be the problem. The problem appears to be related to the fact that ROLL_REF is not near zero.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564501#comment-564501):_

I note that the value of ROLL_REF=-138.76 does seem to make sense, given that (for these exposures) PA_V3=0 and V3I_YANG=138.76, wth VPARITY=-1. So the roll at the aperture ref point is simply equal to the (negative of) the roll of the NRS focal plane in the V2/V3 frame.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564530#comment-564530):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) If either of you want to try to look into this, example datasets are in ███████████████████████████████████████████ There are 3 sets of files: "old" refers to the exposures using the old ROLL_REF value of 359.8, "new" refers to the exposures using the new ROLL_REF value of -138.8, and "new2" refers to the exposures using the new ROLL_REF value that's had 360 added to it (to make it positive 221.2). The asn files "jw84700-o001_20211129t070029_spec3_002_asn_[old|new|new2].json" can be used to run the calwebb_spec3 pipeline using the old, new, new2 exposures as input. When you run:

strun calwebb_spec3 <asn_file.json>

you can see in the outlier_detection step already that it's creating a median image of ridiculous size when processing either the "new" or "new2" exposures.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564624#comment-564624):_

I proposed a fix in #6533

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2340?focusedCommentId=564974#comment-564974):_

Fixed by #6533
